core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -0,0 +1,178 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import scala.collection.mutable
+import scala.concurrent.duration.FiniteDuration
+
+import akka.actor.{PoisonPill, ActorRef, Actor}
+
+import org.apache.spark.{SparkConf, Logging}
+import org.apache.spark.util.{AkkaUtils, ActorLogReceive}
+
+private[spark] sealed trait OutputCommitCoordinationMessage extends Serializable
+
+private[spark] case class StageStarted(stage: Int) extends OutputCommitCoordinationMessage
+private[spark] case class StageEnded(stage: Int) extends OutputCommitCoordinationMessage
+private[spark] case object StopCoordinator extends OutputCommitCoordinationMessage
+
+private[spark] case class AskPermissionToCommitOutput(
+    stage: Int,
+    task: Long,
+    taskAttempt: Long)
+    extends OutputCommitCoordinationMessage
+
+private[spark] case class TaskCompleted(
+    stage: Int,
+    task: Long,
+    attempt: Long,
+    successful: Boolean)
+    extends OutputCommitCoordinationMessage
+
+/**
+ * Authority that decides whether tasks can commit output to HDFS.
+ *
+ * This lives on the driver, but the actor allows the tasks that commit
+ * to Hadoop to invoke it.
+ */
+private[spark] class OutputCommitCoordinator(conf: SparkConf) extends Logging {
+
+  // Initialized by SparkEnv
+  var coordinatorActor: Option[ActorRef] = None
+  private val timeout = AkkaUtils.askTimeout(conf)
+  private val maxAttempts = AkkaUtils.numRetries(conf)
+  private val retryInterval = AkkaUtils.retryWaitMs(conf)
+
+  private type StageId = Int
+  private type TaskId = Long
+  private type TaskAttemptId = Long
+
+  private val authorizedCommittersByStage:
+  mutable.Map[StageId, mutable.Map[TaskId, TaskAttemptId]] = mutable.HashMap()
+
+  def stageStart(stage: StageId) {
+    sendToActor(StageStarted(stage))
+  }
+  def stageEnd(stage: StageId) {
+    sendToActor(StageEnded(stage))
+  }
+
+  def canCommit(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId): Boolean = {
+    askActor(AskPermissionToCommitOutput(stage, task, attempt))
+  }
+
+  def taskCompleted(
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId,
+      successful: Boolean) {
+    sendToActor(TaskCompleted(stage, task, attempt, successful))
+  }
+
+  def stop() {
```

Comment:
Minor, but I think it's slightly weird that this class mixes methods that should only be called from the driver (such as `stop`) and methods that executors can call safely. Perhaps a check here that this is only being called on the driver side?